test/distributed/test_ddp_with_rpc.py

```diff
@@ -0,0 +1,541 @@
+        rank: int,
+    ):
+        self.process_group_context = ProcessGroupContext(
+            name=TRAINER_NAME_TEMPLATE.format(rank),
+            ranks=DDP_TRAINER_RANKS,
+            group_name="trainers_ddp",
+        )
+        process_group_for_ddp = self.process_group_context.__enter__()
+        check_process_group_type(f"trainer #{rank}", process_group_for_ddp)
+        gLogger.info(
+            f"Initing a trainer with process group {process_group_for_ddp} ..."
+        )
+
+        self.criterion = torch.nn.CrossEntropyLoss()
+        self.remote_em_rref = remote_em_rref
+        self.remote_net_rref = remote_net_rref
+        self.hybrid_module = HybridModel(
+            self.remote_em_rref,
+            self.remote_net_rref,
+            process_group_for_ddp if ddp_mode in (DdpMode.INSIDE,) else None,
+        )
+        if ddp_mode == DdpMode.OUTSIDE:
+            gLogger.info(f"Wrapping the whole hybride module into DDP.")
+            self.hybrid_module = DistributedDataParallel(
+                self.hybrid_module,
+                process_group=process_group_for_ddp,
+                check_reduction=True,
+            )
+        self.local_optimizer = optim.SGD(self.hybrid_module.parameters(), lr=LR)
+        gLogger.info(f"Succeeded in creating a HybridModel instance.")
+
+    def __del__(self):
+        self.process_group_context.__exit__(None, None, None)
+
+    def do_forward_without_grad(self, input: FeatureSet):
+        gLogger.debug(f"Doing a forward pass on {input}")
+        with torch.no_grad():
+            output = self.hybrid_module(input)
+            return self.criterion(output, input.labels)
+
+    def get_real_module(self):
+        if isinstance(self.hybrid_module, DistributedDataParallel):
+            return self.hybrid_module.module
+        else:
+            return self.hybrid_module
+
+    def print_parameters(self):
+        self.get_real_module().print_parameters()
+
+    def do_mini_batch(self, mini_batch: FeatureSet):
+        gLogger.info(f"Doing a mini batch on {mini_batch}")
+        loss = 0
+
+        with dist_autograd.context() as context_id:
+            self.local_optimizer.zero_grad()
+            output = self.hybrid_module.forward(mini_batch)
+            with torch.no_grad():
+                gLogger.debug(
+                    f"Output: {output} softmax: {torch.softmax(output, 0)} labels: {mini_batch.labels}"
+                )
+            loss = self.criterion(output, mini_batch.labels)
+            # grads will be stored in dist_autograd context
+            dist_autograd.backward([loss])
+            gLogger.debug(
+                f"Distributed grads: {dist_autograd.get_gradients(context_id)}"
+            )
+            # The local optimize is unable to update the parameters
+            self.local_optimizer.step()
+            self.print_parameters()
+
+            gLogger.info(f"Running distributed optimizer...")
+            dist_optimizer = DistributedOptimizer(
```

Comment:
nit: Typically the optimizer is initialized at the beginning (before we start training). For example if we had a training loop here, we wouldn't initialize the optimizer in each iteration.

Reply:
I see. I thought `DistributedOptimizer` needs to be in the context while being created.